### ISSUE TITLE
symlink breaks docker build: "Cannot locate specified Dockerfile: Dockerfile"

### ISSUE THREAD

**Issue by pwaller** [docker/docker#12842]
If the path argument to docker build is a symlink, it can fail with the error message "Cannot locate specified Dockerfile: Dockerfile".

In the example below I make a directory called `bar` with a `Dockerfile` and then make a symlink `baz -> bar`. If I go to `baz` and run `docker build .` it works fine, but if I specify `docker build $PWD`, it breaks.

Steps to reproduce, docker 1.6.0:

```
~/tmp $ mkdir -p foo/bar

~/tmp $ cd foo/bar/

~/tmp/foo/bar $ echo FROM scratch > Dockerfile

~/tmp/foo/bar $ cd ..

~/tmp/foo $ ln -s bar baz

~/tmp/foo $ cd baz

~/tmp/foo/baz $ docker build .
Sending build context to Docker daemon 2.048 kB
Sending build context to Docker daemon 
Step 0 : FROM scratch
 ---> 
INFO[0000] No image was generated. Is your Dockerfile empty? 

~/tmp/foo/baz $ pwd
/home/pwaller/tmp/foo/baz

~/tmp/foo/baz $ docker build $PWD
Sending build context to Docker daemon 1.536 kB
Sending build context to Docker daemon 
FATA[0000] Error response from daemon: Cannot locate specified Dockerfile: Dockerfile 

~/tmp/foo/baz $ ls -la $PWD
total 4
drwxrwxr-x 1 pwaller pwaller 20 Apr 28 10:06 .
drwxrwxr-x 1 pwaller pwaller 12 Apr 28 10:06 ..
-rw-rw-r-- 1 pwaller pwaller 13 Apr 28 10:06 Dockerfile
```

**Comment by vdemeester** [comment]
It defenitely happens if you're using the absolute path. Seems like a "followSymlink" bug :P

**Comment by pwaller** [comment]
[I instrumented here](https://github.com/docker/docker/blob/779939e334ea33081d3ad7a4364be1b1b4801c3d/pkg/archive/archive.go#L472) with:

```
logrus.Warnf("Tarring %s -> %s", filePath, relFilePath)
```

And the output is as so:

```
WARN[0000] Tarring /home/pwaller/tmp/foo/baz -> .       
```

**Comment by brahmaroutu** [comment]
#dibs
I have addressed similar issue before to make symlinks work for directories. #8420
Do you think the following implementation works for you?

**Comment by pwaller** [comment]
Hmm. I'm not sure the implementation in 1a1737e is as desired. If I understand, that makes the build context follow all symlinks which sounds like a sweeping change with potentially wide implications (e.g, security). In the first instance, I think the problem in this bug is that the behaviour of `docker build $PWD` and `docker build .` can be different.

The question of whether docker follows all symlinks outside the context sounds like a separate issue where I imagine lots of people might have a say on how that should behave.

(Apologies if I misunderstood something).

**Comment by brahmaroutu** [comment]
I Apologize. You are right, I was playing with something in house for allowing ADD and COPY to work with files outside build context. 
Part of the code in the solution can be applied here to make this work. I will give it a try.

**Comment by thaJeztah** [comment]
Looks a bit similar to https://github.com/docker/docker/issues/7804 (and looks like that one shouldn't have been closed)

**Comment by duglin** [comment]
> In the first instance, I think the problem in this bug is that the behaviour of docker build $PWD and docker build . can be different.

I think they may be related because I think what I'm seeing is that in the `$PWD` case the context is sent over to the daemon but because the entire dir is a symlink its ignored - thus it can't find the Dockerfile.
In the `.` case, I think, we never give the "ignore symlink" code a chance to ignore `.`.

**Comment by pwaller** [comment]
I think this was fixed by #15039 @thaJeztah, thanks for the link.

**Comment by thaJeztah** [comment]
Thanks for confirming, @pwaller !
